internal/own/codeowners/proto/match.go

```diff
@@ -0,0 +1,196 @@
+package proto
+
+import (
+	"fmt"
+	"math/big"
+	"strings"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// FindOwners returns the Owners associated with given path as per this CODEOWNERS file.
+// Rules are evaluated in order: Returned owners come from the rule which pattern matches
+// given path, that is the furthest down the file.
+func (f *File) FindOwners(path string) []*Owner {
+	var owners []*Owner
+	for _, r := range f.GetRule() {
+		m, err := compile(r.GetPattern())
+		if err != nil {
+			continue
+		}
+		if m.match(path) {
+			owners = r.GetOwner()
+		}
+	}
+	return owners
+}
+
+const separator = "/"
+
+// globPattern implements a pattern for matching file paths,
+// which can use directory/file names, * and ** wildcards,
+// and may or may not be anchored to the root directory.
+type globPattern []patternPart
+
+// compile translates a text representation of a glob pattern
+// to an executable one that can `match` file paths.
+func compile(pattern string) (globPattern, error) {
+	var p []patternPart
+	// No leading `/` is equivalent to prefixing with `/**/`.
+	// The pattern matches arbitrarily down the directory tree.
+	if !strings.HasPrefix(pattern, separator) {
+		p = append(p, anySubPath{})
+	}
+	for _, part := range strings.Split(strings.Trim(pattern, separator), separator) {
+		switch {
+		case part == "":
+			return nil, errors.New("two consecutive forward slashes")
+		case part == "*":
+			p = append(p, anyMatch{})
+		case part == "**":
+			p = append(p, anySubPath{})
+		default:
+			p = append(p, exactMatch(part))
+		}
+	}
```

Comment:
You could rewrite this to remove the `part ==` from every case:

```suggestion
		switch part {
		case "":
			return nil, errors.New("two consecutive forward slashes")
		case "*":
			p = append(p, anyMatch{})
		case "**":
			p = append(p, anySubPath{})
		default:
			p = append(p, exactMatch(part))
		}
	}
```